### DISCUSSION TITLE
DeleteSelf

### DISCUSSION THREAD

**Discussion by Brad-IMS** [MarimerLLC/csla#3440] [General]
I successfully upgraded from CSLA 4.5.1 to CSLA 7.0 several months ago.  However, I recently had occasion to perform a self-delete on a record for the first time and found that there didn't seem to be any way to do it.  I see that in 4.5.1 DeleteSelf was designated as protected, whereas in the 7.0 templates and snippets it's shown as private.  Is this an error, or is there an alternative way of invoking it in 7.0?

**Comment by rockfordlhotka** [comment]
As of CSLA 5 the data portal no longer relies on specific method names, and instead relies on attributes such as `[Fetch]` to identify the method to be invoked. There's an attribute for the delete self operation.

The data portal operation methods do need to be private, and now can have any name you choose.

Starting in CSLA 6 and 7 we've been phasing out the backward support for the legacy "DataPortal_XYZ" method names. Eventually the attributes will be the required approach.

**Reply by Brad-IMS** [reply]
I based my implementation on BlazorExample in the Samples directory.  BlazorExample has this code in the PersonEdit class:

```
    [DeleteSelf]
    private void DeleteSelf([Inject]DataAccess.IPersonDal dal)
    {
      Delete(ReadProperty(IdProperty), dal);
    }

    [Delete]
    private void Delete(int id, [Inject]DataAccess.IPersonDal dal)
    {
      dal.Delete(id);
    }
```
And this in PersonEFDal:

```
    public bool Delete(int id)
    {
      var person = _context.Persons.Where(p => p.Id == id).FirstOrDefault();
      if (person != null)
      {
        _context.Persons.Remove(person);
        _context.SaveChanges();     
        return true;
      }
      else
      {
        return false;
      }
    }
```

personEditPortal only gives a choice of Delete or DeleteAsync.  How does it know whether to invoke Delete or DeleteSelf?  Should there also be a Delete() in PersonEFDal?

Also, shouldn't the [Delete] example be 'bool deleted = dal.Delete(id);' so you know whether or not it succeeded?

**Reply by rockfordlhotka** [reply]
The rules for how delete and deleteself work are the same as always.

The data portal `Delete` and `DeleteAsync` methods invoke the deleteself operation method.

When you call `SaveAsync` on a root object, if objects in the graph are marked for deletion then the delete operation method is invoked.

**Reply by Brad-IMS** [reply]
OK.  I was calling objPortal.DeleteAsync() when I should have been calling obj.Delete() and obj.SaveAsync().
